### ISSUE TITLE
NavigationPrompt on window close

### ISSUE THREAD

**Issue by calebmer** [ReactTraining/react-router#4272]
Currently it looks like using a `<NavigationPrompt>` component in `react-router` v4 does not prompt when the user is trying to close the tab like Codepen does with an [unsaved pen](http://codepen.io/pen/). Is this intentional behavior? If not, I’ll put together a test case.

**Comment by pshrmn** [comment]
This should be handled in `history`. v2/3 of `history` include a [`useBeforeUnload`](https://github.com/mjackson/history/blob/v3.2.1/modules/useBeforeUnload.js) history enhancer, but it doesn't look like `beforeunload` support was included in the v4 rewrite. Probably best to move this issue over there. Once it is supported by `history` it would just be a matter of adding is as a prop to the `<__Router>`s.

**Comment by mjackson** [comment]
I was thinking about supporting this use case with [a `beforeUnload` prop to `<NavigationPrompt>`](https://github.com/mjackson/react-history/issues/3).

**Comment by gpltaylor** [comment]
@mjackson what would the prop type be? function / Promise / bool / string
I would like the choice of two different messages, one for simple navigation away, and another for leaving the application.

Example:
` <NavigationPrompt`
` when={this.state.blockTransitions}`
`message={(location) => ("Are you sure you want to go to ${location.pathname}")}`
`beforeUnload={(location) => ("Are you sure you want to leave")} />`

Would we check to make sure if window.onbeforeunload is already subscribed?

Would this need to be added to "History" and used within "React-Router" like the existing blockTransitions?

`blockTransitions={history.block},`
`onBeforeUnload={history.unload}`

Would we honour both (when/beforeUnload) when the page is closed, or just beforeUnload?

**Comment by gpltaylor** [comment]
Just found out that you **can't** provide a custom message for the event (onbeforeunload)! Every day is a school day..

`window.onbeforeunload`

For this reason should beforeUnload be a boolean?

```
<NavigationPrompt
when={this.state.blockTransitions}
message={(location) => ("Are you sure you want to go to ${location.pathname}")}
beforeUnload={true} />
```

**Comment by gpltaylor** [comment]
Rethink, beforeUnload should be of type (bool/function). This will allow a simple (use default browser behaviour) or custom action.

For example, if we need to do some house-keeping before asking if they wish to leave.

usage as boolean
```
<NavigationPrompt
when={this.state.blockTransitions}
message={(location) => ("Are you sure you want to go to ${location.pathname}")}
beforeUnload={true} />
```

usage as function
```
<NavigationPrompt
when={this.state.blockTransitions}
message={(location) => ("Are you sure you want to go to ${location.pathname}")}
beforeUnload={() => { console.log("my custom clean up code"); return "leave"; }} />
```

I have a VERY basic Gist. Please, let me know if you like the direction and what needs improvement.
https://gist.github.com/gpltaylor/80e932c859d5e2dd1471555e610fe994

**Comment by gavinwahl** [comment]
I also expected this behavior.

> Just found out that you can't provide a custom message for the event (onbeforeunload)!

I thought you could, but some browsers ignore it?

**Comment by gpltaylor** [comment]
@gavinwahl browsers have been updated to use generic messages.

Starting with Firefox 4, Chrome 51, Opera 38 and Safari 9.1, a generic string not under the control of the webpage will be shown instead of the returned string.

https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload

**Comment by gpltaylor** [comment]
@gavinwahl if you want to test this I have a fork (https://github.com/gpltaylor/react-router)

**Comment by mjackson** [comment]
@gpltaylor v4 is settling down now. If you'd like to take a look at adding this behavior to our `<Prompt>` component I'd be grateful.

**Comment by gpltaylor** [comment]
@mjackson I would love too. I am doing a talk in Manchester about React Router and I'm currently writing a book about the latest version, so it would be great to contribute to the code base. I will PR when ready.

**Comment by mjackson** [comment]
Let's follow up this discussion in #4359

**Comment by srirammanohar** [comment]
Which version is this fixed in? I dont see the fix in react 4.2.2
